File: Aegis/Assets/Scripts/Player/PlayerStats.js
#pragma strict

var heat : float = 50;
var maxHeat : float = 100;
var heatCooldownRate : float = 1; 		
var overheat : boolean = false;
var overheatTime : float = 1;			//time you are overheated
var speedReduction : float = 0.5;		//overheat penalty
var expWindow : float = 2;

public static var totalCredits : int  = 5000;
var creditsThisLevel : int = 0;
private var beginOverheat : float;
var levelUp : AudioClip; 

enum WeaponType {
	Projectile,
	Explosive,
	Energy,
	Device
}
class WeaponLevel extends System.Object{
	var damage : float;
	var heatCost : float;
	var cooldown : float;
	
	@System.NonSerialized
	var lastShot : float;

}

class Weapon extends System.Object {
	var level : int = 0;
	var name : String;
	var level0 : WeaponLevel;
	var level1 : WeaponLevel;
	var level2 : WeaponLevel;
	var level3 : WeaponLevel;
	var weaponType : WeaponType = weaponType.Projectile;
	@System.NonSerialized
	var input : String;
	@System.NonSerialized
	var currentLevel : WeaponLevel;
	@System.NonSerialized
	var currentExp : float;
}

var allEquipables = new Array();
var weapons : Weapon[];
selections = new String[4];
var VulcanCannonStats : Weapon;
var FusionBeamStats : Weapon;
var ShieldStats : Weapon;
var CoolantTankStats : Weapon;
var selections : String[];


function Awake () {
	if (Application.platform != RuntimePlatform.WindowsEditor){
		selections = UpgroidsGUI.selections.ToBuiltin(String) as String[];
	}
	var allEquipables = new Array(VulcanCannonStats, FusionBeamStats,ShieldStats,CoolantTankStats);
	var convAllEquipables :Weapon[] = allEquipables.ToBuiltin(Weapon) as Weapon[];
	weapons = new Weapon[4];
	var weaponsIndex : int = 0;
	for (var i : int = 0; i < selections.length; i++){
		for (var k : int = 0; k < convAllEquipables.length; k++){
			if (selections[i] == convAllEquipables[k].name){
				weapons[weaponsIndex] = convAllEquipables[k];
				var str = convAllEquipables[k].name.Replace(' ','');
				gameObject.AddComponent(""+str);
				weaponsIndex++;
			}
		}
	}
	weapons[0].input = InputCoordinator.weaponOne;
	weapons[1].input = InputCoordinator.weaponTwo;
	weapons[2].input = InputCoordinator.weaponThree;
	weapons[3].input = InputCoordinator.weaponFour;
	CheckLevel();
}

function Update () {
	CheckLevel();
	if (heat > 0 && heat != maxHeat){
		heat = Mathf.MoveTowards(heat, 0, heatCooldownRate * Time.deltaTime);
	}
	if (heat > maxHeat){ //overheated!
		heat = maxHeat;
		overheat = true;
		beginOverheat = Time.time;
		gameObject.GetComponent(Stats).speed *= speedReduction;
	}
	if (overheat && Time.time - beginOverheat >= overheatTime){ //end overheat
		overheat = false;
		heat = 99;
		
		gameObject.GetComponent(Stats).speed /= speedReduction;
	}
}

function CheckLevel () {
	for (var weapon : Weapon in weapons){
		if (weapon.currentExp >= 100){
			weapon.level += 1;
			weapon.currentExp -= 100;
			Debug.Log(weapon.name+" levelled up!");
			audio.PlayOneShot(levelUp, 1);
			Camera.main.GetComponent(MyGUI).levelledName = weapon.name;
			Camera.main.GetComponent(MyGUI).levelledLevel = weapon.level;
			Camera.main.GetComponent(MyGUI).levelledTime = weapon.currentLevel.lastShot;
			Camera.main.GetComponent(MyGUI).boolLevel = true;
		}
	}
	if (VulcanCannonStats.level == 0)VulcanCannonStats.currentLevel = VulcanCannonStats.level0;
	if (VulcanCannonStats.level == 1)VulcanCannonStats.currentLevel = VulcanCannonStats.level1;
	if (VulcanCannonStats.level == 2)VulcanCannonStats.currentLevel = VulcanCannonStats.level2;
	if (VulcanCannonStats.level == 3)VulcanCannonStats.currentLevel = VulcanCannonStats.level3;
	
	if (FusionBeamStats.level == 0)FusionBeamStats.currentLevel = FusionBeamStats.level0;
	if (FusionBeamStats.level == 1)FusionBeamStats.currentLevel = FusionBeamStats.level1;
	if (FusionBeamStats.level == 2)FusionBeamStats.currentLevel = FusionBeamStats.level2;
	if (FusionBeamStats.level == 3)FusionBeamStats.currentLevel = FusionBeamStats.level3;
	
	if (ShieldStats.level == 0)ShieldStats.currentLevel = ShieldStats.level0;
	if (ShieldStats.level == 1)ShieldStats.currentLevel = ShieldStats.level1;
	if (ShieldStats.level == 2)ShieldStats.currentLevel = ShieldStats.level2;
	if (ShieldStats.level == 3)ShieldStats.currentLevel = ShieldStats.level3;
	
	if (CoolantTankStats.level == 0)CoolantTankStats.currentLevel = CoolantTankStats.level0;
	if (CoolantTankStats.level == 1)CoolantTankStats.currentLevel = CoolantTankStats.level1;
	if (CoolantTankStats.level == 2)CoolantTankStats.currentLevel = CoolantTankStats.level2;
	if (CoolantTankStats.level == 3)CoolantTankStats.currentLevel = CoolantTankStats.level3;
}

function GiveExp (exp : float, time : float){
	for (var weapon : Weapon in weapons){
		if (Time.time - weapon.currentLevel.lastShot < expWindow){
			weapon.currentExp += exp;
		}
	}
}

File: Aegis/Assets/Scripts/Player/PlayerStats.cs
using UnityEngine;
using System.Collections;

public class PlayerStats : MonoBehaviour, IKillable, IDamageable<float>, ICollisionDamage, ISpeed
{
    [SerializeField]
    private float speed;
    public float Speed
    {
        get { return speed; }
        set { speed = value; }
    }
    [SerializeField]
    private float health;
    public float Health
    {
        get { return health; }
        set { health = value; }
    }
    [SerializeField]
    private float collisionDamage;
    public float CollisionDamage
    {
        get { return collisionDamage; }
        set { collisionDamage = value; }
    }
    public float maxHealth;
    public float heat = 10;
    public float maxHeat = 100;
    public float heatCooldownRate = 1;
    public bool overheat = false;
    public float overheatTime = 1;
    float ltOverheat = 0;
    public float speedReduction = 0.5f;

    public float expWindow = 2;

    public bool invincible = false;

    public static float totalCredit = 5000; //convert to playerPrefs
    public float creditsThisLevel = 0;

    public AudioClip levelUp;

	// Update is called once per frame
	void Update () 
    {
        if (heat > 0 && heat != maxHeat) //cooldown
            heat = Mathf.MoveTowards(heat, 0, heatCooldownRate * Time.deltaTime);
        if (heat > maxHeat)             //overheated!
        {
            heat = maxHeat;
            overheat = true;
            ltOverheat = Time.time;
            speed *= speedReduction;
        }
        if (overheat && Time.time - ltOverheat >= overheatTime)
        {
            overheat = false;
            heat = 99.9f;
            speed /= speedReduction;
        }
        if (invincible)
            health = 100;
        if (health <= 0)
        {
            Kill();
        }
	}

    public void Damage(float damage)
    {
        Health = Health - damage;
    }
    public void GiveExp (float exp)
    {
        float[] times = new float[3];
        times[0] = GetComponent<BlunderBuster>().blunderBuster.currentLevel.ltShot;
        times[1] = GetComponent<FusionBeam>().fusionBeam.currentLevel.ltShot;
        times[2] = GetComponent<ArcNade>().arcNade.currentLevel.ltShot;
        float greatest = Mathf.Max(times);
        if (greatest == times[0])
        {
            GetComponent<BlunderBuster>().blunderBuster.GiveExp(exp);
            GetComponent<BlunderBuster>().blunderBuster.CheckForLevelUp(Camera.main.gameObject);
        }
        if (greatest == times[1])
        {
            GetComponent<FusionBeam>().fusionBeam.GiveExp(exp);
            GetComponent<FusionBeam>().fusionBeam.CheckForLevelUp(Camera.main.gameObject);
        }
        if (greatest == times[2])
        {
            GetComponent<ArcNade>().arcNade.GiveExp(exp);
            GetComponent<ArcNade>().arcNade.CheckForLevelUp(Camera.main.gameObject);
        }
    }
    public void Kill()
    {
        //GetComponent<GameOverlayGUI>().lost = true;
        Destroy(gameObject);
    }

    public void OnCollisionEnter (Collision other)
    {
        other.gameObject.SendMessage("Damage", CollisionDamage);
    }

    public void DamageProjectile(float damageTaken)
    {
        Damage(damageTaken);
    }

    public void DamageExplosive(float damageTaken)
    {
        Damage(damageTaken);
    }

    public void DamageEnergy(float damageTaken)
    {
        Damage(damageTaken);
    }

    public void SuperEffectiveSystem()
    {
    }
}
